pkg/yurtctl/cmd/join/join.go

```diff
@@ -0,0 +1,399 @@
+	externalcfg.ControlPlane = &kubeadmapiv1beta2.JoinControlPlane{}
+
+	// Apply defaults
+	kubeadmscheme.Scheme.Default(externalcfg)
+
+	return &joinOptions{
+		externalcfg: externalcfg,
+	}
+}
+
+// NewJoinCmd returns "yurtctl join" command.
+func NewCmdJoin(out io.Writer, joinOptions *joinOptions) *cobra.Command {
+	if joinOptions == nil {
+		joinOptions = newJoinOptions()
+	}
+	joinRunner := workflow.NewRunner()
+
+	cmd := &cobra.Command{
+		Use:   "join [api-server-endpoint]",
+		Short: "Run this on any machine you wish to join an existing cluster",
+		RunE: func(cmd *cobra.Command, args []string) error {
+			c, err := joinRunner.InitData(args)
+			if err != nil {
+				return err
+			}
+			data := c.(*joinData)
+			if err := joinRunner.Run(args); err != nil {
+				return err
+			}
+			fmt.Fprint(data.outputWriter, joinWorkerNodeDoneMsg)
+			return nil
+		},
+	}
+
+	addJoinConfigFlags(cmd.Flags(), joinOptions)
+
+	joinRunner.AppendPhase(yurtphase.NewPreparePhase())
+	joinRunner.AppendPhase(kubeadmPhase.NewPreflightPhase())
+	joinRunner.AppendPhase(yurtphase.NewEdgeNodePhase())
+	joinRunner.AppendPhase(yurtphase.NewCloudNodePhase())
+	joinRunner.AppendPhase(yurtphase.NewPostcheckPhase())
+
+	joinRunner.SetDataInitializer(func(cmd *cobra.Command, args []string) (workflow.RunData, error) {
+		return newJoinData(cmd, args, joinOptions, out)
+	})
+	joinRunner.BindToCommand(cmd)
+	return cmd
+}
+
+// addJoinConfigFlags adds join flags bound to the config to the specified flagset
+func addJoinConfigFlags(flagSet *flag.FlagSet, joinOptions *joinOptions) {
+	flagSet.StringVar(
+		&joinOptions.externalcfg.NodeRegistration.Name, options.NodeName, joinOptions.externalcfg.NodeRegistration.Name,
+		`Specify the node name.`,
+	)
+	flagSet.StringSliceVar(
+		&joinOptions.externalcfg.Discovery.BootstrapToken.CACertHashes, options.TokenDiscoveryCAHash, []string{},
+		"For token-based discovery, validate that the root CA public key matches this hash (format: \"<type>:<value>\").",
+	)
+	flagSet.BoolVar(
+		&joinOptions.externalcfg.Discovery.BootstrapToken.UnsafeSkipCAVerification, options.TokenDiscoverySkipCAHash, false,
+		"For token-based discovery, allow joining without --discovery-token-ca-cert-hash pinning.",
+	)
+	flagSet.StringSliceVar(
+		&joinOptions.ignorePreflightErrors, options.IgnorePreflightErrors, joinOptions.ignorePreflightErrors,
+		"A list of checks whose errors will be shown as warnings. Example: 'IsPrivilegedUser,Swap'. Value 'all' ignores errors from all checks.",
+	)
+	flagSet.StringVar(
+		&joinOptions.token, options.TokenStr, "",
+		"Use this token for both discovery-token and tls-bootstrap-token when those values are not provided.",
+	)
+	flagSet.StringVar(
+		&joinOptions.nodeType, "node-type", "",
+		"Sets the node is edge-node or cloud-node",
+	)
+	flagSet.StringVar(
+		&joinOptions.yurthubImage, "yurthub-image", "",
+		"Sets the node is edge-node or cloud-node",
```

Comment:
--> Sets the image version of yurthub component

Reply:
done